cpp/arcticdb/vector_db/vector_db.cpp

```diff
@@ -0,0 +1,102 @@
+                                    this->query_vector_.size());
+                            col->type().visit_tag([&top_k, &slice_and_key, &store, &col=col, idx=idx, lp, this](auto type_desc_tag) {
+                                using TypeDescriptorTag = decltype(type_desc_tag);
+                                using RawType = typename TypeDescriptorTag::DataTypeTag::raw_type;
+
+                                if constexpr (is_floating_point_type(TypeDescriptorTag::DataTypeTag::data_type)) {
+
+                                    ColumnData col_data = col->data();
+                                    double sum_differences_exponentiated = 0;
+                                    auto j = 0u;
+
+                                    while (auto block = col_data.next<TypeDescriptorTag>()) {
+                                        auto ptr = reinterpret_cast<const RawType *>(block.value().data());
+                                        for (auto progress_in_block = 0u; progress_in_block <
+                                                                          block.value().row_count(); ++progress_in_block, ++j, ++ptr) {
+                                            sum_differences_exponentiated += pow(
+                                                    abs(*ptr - this->query_vector_[j]), lp);
+                                        }
+                                    }
+
+                                    top_k.try_insert(NamedColumnSortedByDistance(
+                                            sum_differences_exponentiated,
+                                            col,
+                                            slice_and_key.segment(store).field(idx).name()));
+                                }
+                            });
+                        }
+                    }
+                }
+        );
+        // By this time we should have iterated through each block of each SegmentInMemory of each ProcessingSegment;
+        // and in each case read a vector and inserted it if more similar than the running top_k_. Now we write to a new
+        // segment.
+        SegmentInMemory seg;
+        seg.descriptor().set_index(IndexDescriptor(0, IndexDescriptor::ROWCOUNT));
+        seg.set_row_id(query_vector_.size() - 1 + 1);
+        // -1 because of zero-indexing, +1 because we want to add the distance on at the end.
+            for (const auto &column: top_k.top_k_) {
+            // todo: replace (mem)cpying of columns' contents with holding of pointers in top_k_ and addition of pointers to seg
+            // this doesn't work presently because the columns can have >1 block.
+            column.column_->type().visit_tag([&column, &seg, lp, this](auto type_desc_tag) {
+                using TypeDescriptorTag = decltype(type_desc_tag);
+                using RawType = typename TypeDescriptorTag::DataTypeTag::raw_type;
+
+                if constexpr (is_floating_point_type(TypeDescriptorTag::DataTypeTag::data_type)) {
```

Comment:
Do we want to add an exception when the column is of some other type, or is silently skipping them okay?

Reply:
Alex suggested this as a compiler optimisation I think? This is already checked in Python.

Reply:
Same as above, add the else to throw so it's super explicit